src/Microsoft.DocAsCode.Build.Engine/XrefServiceResolver.cs

```diff
@@ -81,8 +82,73 @@ public XrefServiceResolver(HttpClient client, ImmutableArray<string> xrefService
 
         private IUriTemplatePipeline<Task<List<XRefSpec>>> GetPipeline(string name)
         {
-            // todo : add pipeline.
-            return EmptyUriTemplatePipeline.Default;
+            // todo: pluggable.
+            switch (name)
+            {
+                case "removeHost":
+                    return RemoveHostUriTemplatePipeline.Default;
+                case "addQueryString":
+                    return AddQueryStringUriTemplatePipeline.Default;
+                default:
+                    Logger.LogWarning($"Unknown uri template pipeline: {name}.", code: WarningCodes.Build.UnknownUriTemplatePipeline);
+                    return EmptyUriTemplatePipeline.Default;
+            }
+        }
+
+        private sealed class RemoveHostUriTemplatePipeline : IUriTemplatePipeline<Task<List<XRefSpec>>>
+        {
+            public static readonly RemoveHostUriTemplatePipeline Default = new RemoveHostUriTemplatePipeline();
+
+            public async Task<List<XRefSpec>> Handle(Task<List<XRefSpec>> value, string[] parameters)
+            {
+                var list = await value;
+                foreach (var item in list)
+                {
+                    if (string.IsNullOrEmpty(item.Href))
+                    {
+                        continue;
+                    }
+                    if (Uri.TryCreate(item.Href, UriKind.Absolute, out var uri))
+                    {
+                        item.Href = item.Href.Remove(uri.GetLeftPart(UriPartial.Path).Length);
+                    }
+                }
+                return list;
+            }
+        }
+
+        private sealed class AddQueryStringUriTemplatePipeline : IUriTemplatePipeline<Task<List<XRefSpec>>>
+        {
+            public static readonly AddQueryStringUriTemplatePipeline Default = new AddQueryStringUriTemplatePipeline();
+
+            public Task<List<XRefSpec>> Handle(Task<List<XRefSpec>> value, string[] parameters)
+            {
+                if (parameters.Length == 2 &&
+                    !string.IsNullOrEmpty(parameters[0]) &&
+                    !string.IsNullOrEmpty(parameters[1]))
+                {
+                    return HandleCoreAsync(value, parameters[0], parameters[1]);
+                }
+                return value;
```

Comment:
>value [](start = 23, length = 5)

any compile step to log warnings for invalid parameters?

Reply:
not enough information in interface.